packages/console/src/pages/ApplicationDetails/components/MachineToMachineApplicationRoles/index.tsx

```diff
@@ -0,0 +1,183 @@
+import type { Application, Role } from '@logto/schemas';
+import { RoleType } from '@logto/schemas';
+import { conditional } from '@silverhand/essentials';
+import { useState } from 'react';
+import { toast } from 'react-hot-toast';
+import { useTranslation } from 'react-i18next';
+import useSWR from 'swr';
+
+import Delete from '@/assets/icons/delete.svg';
+import Plus from '@/assets/icons/plus.svg';
+import EmptyDataPlaceholder from '@/components/EmptyDataPlaceholder';
+import { defaultPageSize } from '@/consts';
+import Button from '@/ds-components/Button';
+import ConfirmModal from '@/ds-components/ConfirmModal';
+import IconButton from '@/ds-components/IconButton';
+import Search from '@/ds-components/Search';
+import Table from '@/ds-components/Table';
+import TextLink from '@/ds-components/TextLink';
+import { Tooltip } from '@/ds-components/Tip';
+import type { RequestError } from '@/hooks/use-api';
+import useApi from '@/hooks/use-api';
+import useSearchParametersWatcher from '@/hooks/use-search-parameters-watcher';
+import AssignToRoleModal from '@/pages/Roles/components/AssignToRoleModal';
+import { buildUrl, formatSearchKeyword } from '@/utils/url';
+
+import * as styles from './index.module.scss';
+
+const pageSize = defaultPageSize;
+
+type Props = {
+  application: Application;
+};
+
+function MachineToMachineApplicationRoles({ application }: Props) {
+  const { t } = useTranslation(undefined, { keyPrefix: 'admin_console' });
+
+  const [{ page, keyword }, updateSearchParameters] = useSearchParametersWatcher({
+    page: 1,
+    keyword: '',
+  });
+
+  const { data, error, mutate } = useSWR<[Role[], number], RequestError>(
+    buildUrl(`api/applications/${application.id}/roles`, {
+      page: String(page),
+      page_size: String(pageSize),
+      ...conditional(keyword && { search: formatSearchKeyword(keyword) }),
+    })
+  );
+
+  const isLoading = !data && !error;
+
+  const [roles, totalCount] = data ?? [];
+
+  const [isAssignRolesModalOpen, setIsAssignRolesModalOpen] = useState(false);
+  const [roleToBeDeleted, setRoleToBeDeleted] = useState<Role>();
+  const [isDeleting, setIsDeleting] = useState(false);
+
+  const api = useApi();
+
+  const handleDelete = async () => {
+    if (!roleToBeDeleted || isDeleting) {
+      return;
+    }
+    setIsDeleting(true);
+
+    try {
+      await api.delete(`api/applications/${application.id}/roles/${roleToBeDeleted.id}`);
+      toast.success(t('application_details.roles.deleted', { name: roleToBeDeleted.name }));
+      await mutate();
+      setRoleToBeDeleted(undefined);
+    } finally {
+      setIsDeleting(false);
+    }
+  };
+
+  return (
+    <>
+      <Table
```

Comment:
Looks similar to the user roles page, is it possible to extract it as a shared one?